modules/lang-expression/src/main/java/org/elasticsearch/script/expression/GeoLatitudeValueSource.java

```diff
@@ -41,14 +41,21 @@ public DoubleValues getValues(LeafReaderContext leaf, DoubleValues scores) {
         LeafGeoPointFieldData leafData = (LeafGeoPointFieldData) fieldData.load(leaf);
         final MultiGeoPointValues values = leafData.getGeoPointValues();
         return new DoubleValues() {
+
+            int doc = -1;
+
             @Override
             public double doubleValue() throws IOException {
-                return values.nextValue().getLat();
+                if (values.advanceExact(doc)) {
```

Comment:
I don't think this is a correct implementation. if the correct impl has different perf characteristics we need to fix them but lets keep it correct. I don't think it should return 0 if there is nothing to return we should rather throw an exception instead? another option would be to assign the value in `advanceExact` instead of the doc and then return the value only in `doubleValue` and assert that we never do it unless it's actually advanced to this doc?

Reply:
I agree with Simon; we shouldn't be calling advance inside the value method. I'm not sure if we want to pull the value in advance though. I think we can simply stash whether advance returned `true`, to know whether calling the value method should throw an assertion or call `values.nextValue().getLat()`?

Reply:
The difference in performance does seem to be down to how far apart the `advanceExact()` and `nextValue()` calls are, so ideally we want to keep them in the same method (as was the case in the previous `ValueSource` implementation.  I can add an assertion to ensure that `doubleValue()` doesn't get called when we're unpositioned though, that would indeed be an error.